browser/ui/brave_ads/ad_notification_view_factory.cc

```diff
@@ -0,0 +1,21 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/brave_ads/ad_notification_view_factory.h"
+
+#include "brave/browser/ui/brave_ads/text_ad_notification_view.h"
+
+namespace brave_ads {
+
+// static
+AdNotificationView* AdNotificationViewFactory::Create(
```

Comment:
what's the purpose of having a factory for creating this? I'm not seeing a good reason, but also it seems to possibly be hiding the fact that this object has no owner and therefore never gets destroyed. Returning a raw pointer like this is generally not a good idea for that reason imo

Reply:
A factory here is to support different types of ad notifications, right now we only have text ad notifications but in the not to distant future we will support image ad notifications, video ad notifications etc.

The object is created and assigned to `ad_notification_view_` in `AdNotificationPopup::AdNotificationPopup` which is then added to and owned by the view, see `AdNotificationPopup::CreatePopup()`

Reply:
ah right, and the parent view then destroys it? I hate the way that works in upstream

Reply:
but then it seems like a problem that you have a global list of them in `g_ad_notification_popups`? What happens when they get destroyed by the parent view?

Reply:
This was taken from an early version of message center, and g_ad_notification_popups contains its own widget and associated views. I have raised https://github.com/brave/brave-browser/issues/14957 to move this logic to a popup manager.